dash_bio/utils/gene_expression_reader.py

```diff
@@ -1,57 +1,151 @@
+import GEOparse as gp
 import pandas as pd
 import tempfile
 
 
 # pylint: disable=unnecessary-lambda
 
-def parse_tsv(
+def read_soft_file(
         contents='',
         filepath='',
-        row_labels_source=None,
         rows=None,
         columns=None,
-        header_rows=5,
-        header_cols=2
+        return_filtered_data=False
 ):
-
     if len(contents) > 0:
-        with tempfile.NamedTemporaryFile(mode='w+', delete=False) as tf:
+        with tempfile.NamedTemporaryFile(
+                mode='w+', delete=False, suffix='.soft'
+        ) as tf:
             tf.write(contents)
             filepath = tf.name
-    df = pd.read_csv(filepath, sep='\t', skiprows=header_rows-1)
 
-    data = {}
+    geo_file = gp.get_GEO(filepath=filepath, geotype='GDS')
 
-    selected_rows = []
-    selected_cols = []
+    df = geo_file.table
+    df.set_index('ID_REF', inplace=True)
+
+    all_rows = list(df.index.values)
+    all_cols = list(df.columns.values)
+
+    for column in all_cols:
+        if 'GSM' not in column:
+            all_cols.remove(column)
 
-    all_rows = []
-    if (row_labels_source is not None) and (row_labels_source in df.keys().tolist()):
-        all_rows = df[row_labels_source].tolist()
+    if not return_filtered_data:
+        desc = geo_file.metadata
+        subsets = geo_file.subsets
 
-    all_cols = df.keys().tolist()[header_cols:]
-    if (rows is not None) and (columns is not None):
-        for r in rows:
-            if r not in all_rows:
-                continue
-            selected_rows.append(r)
-        selected_rows = list(map(lambda x: all_rows.index(x), selected_rows))
+        for subset in geo_file.subsets:
+            subsets[subset] = geo_file.subsets[subset].metadata
 
-        for c in columns:
-            if c not in all_cols:
-                continue
-            selected_cols.append(c)
+        return desc, subsets, all_rows, all_cols
 
-        selected_data = df.loc[selected_rows, selected_cols]
-        data = selected_data.values
+    return _get_selected_data(
+        df,
+        all_rows,
+        all_cols,
+        rows,
+        columns
+    )
 
+
+def read_tsv_file(
+        contents='',
+        filepath='',
+        rows=None,
+        columns=None,
+        description_row_prefix='#',
+        description_separator=':',
+        skiprows=0,
+        ignore_columns=None,
+        index_column='Gene Name',
+        return_filtered_data=False
+):
     desc = {}
-    info = pd.read_csv(filepath, sep='^', nrows=header_rows-1, header=None)[0]
-    for i in info:
-        tmp = i.strip('#').split(':', 1)
-        if len(tmp) < 2:
-            desc['Source'] = tmp[0]
+
+    if len(contents) > 0:
```

Comment:
We may also want to make sure it's a character string.

Reply:
Let's do this in a subsequent PR. There are actually a few parsers that use this pattern -- e.g., https://github.com/plotly/dash-bio/blob/96d43e90c4598b42282ee27b28e827c8e0f6ec7e/dash_bio/utils/protein_reader.py#L29 
I'd prefer to get those all done in one go, if that's OK!

Reply:
Actually, we'll take care of it as we address #330.